src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -2008,13 +2008,18 @@ void G1ConcurrentMark::print_stats() {
 }
 
 void G1ConcurrentMark::concurrent_cycle_abort() {
-  if (!cm_thread()->in_progress() || _has_aborted) {
-    // We haven't started a concurrent cycle or we have already aborted it. No need to do anything.
+  // We haven't started a concurrent cycle no need to do anything; we might have
+  // aborted the marking because of shutting down. The marking might have already
+  // completed the abort (leading to in_progress() below to return false), however
+  // this left marking state particularly in the shared marking bitmap that must
+  // be cleaned up.
+  // If there are multiple full gcs during shutdown we do too much work.
+  if (!cm_thread()->in_progress() && !_g1h->concurrent_mark_is_terminating()) {
     return;
   }
 
-  // Clear all marks in the next bitmap for the next marking cycle. This will allow us to skip the next
-  // concurrent bitmap clearing.
+  // Clear all marks in the next bitmap for the next marking cycle. This will allow
+  // us to skip the next concurrent bitmap clearing.
```

Comment:
Pre-existing, but since you touch it. This comment could be brushed up even more to be more clear that this clearing is needed for the Full GC as well. I mean the only caller to `concurrent_cycle_abort()` is the Full GC so would not hurt. Not sure how old the comment is, but prior to the parallel Full GC we did not use the bitmap in the collection (but now we do).

Reply:
I touched up these comments a bit. Thanks for noticing.